File: examples/comparison/counter_service.go
// Go: Counter Service
// Same problem as counter_service.hew but in Go.
// Notice: manual channel/mutex management, no supervision,
// no wire types, race conditions possible.
package main

import (
	"fmt"
	"sync"
)

// Go has no actor model. You choose between channels or mutexes.
// Both require discipline. The compiler won't catch data races.

type Counter struct {
	mu    sync.Mutex // You must remember this
	count int32
}

func NewCounter() *Counter {
	return &Counter{}
}

func (c *Counter) Increment(amount int32) int32 {
	c.mu.Lock()         // You must remember this
	defer c.mu.Unlock() // You must remember this
	c.count += amount
	return c.count
}

func (c *Counter) GetCount() int32 {
	c.mu.Lock()         // You must remember this
	defer c.mu.Unlock() // You must remember this
	return c.count
}

// Go has no supervision trees. You build your own restart logic.
type CounterPool struct {
	counters []*Counter
	maxRestarts int
	restartCount int
}

func NewCounterPool(size int) *CounterPool {
	pool := &CounterPool{
		counters:    make([]*Counter, size),
		maxRestarts: 5,
	}
	for i := range pool.counters {
		pool.counters[i] = NewCounter()
	}
	return pool
}

// No language support for restart strategies. You write this yourself.
func (p *CounterPool) RestartChild(idx int) {
	p.restartCount++
	if p.restartCount > p.maxRestarts {
		fmt.Println("Exceeded restart limit")
		return
	}
	p.counters[idx] = NewCounter() // Manual restart
}

// Go has no wire types. You need protobuf + protoc + code generation.
// This is a plain struct. No schema evolution. No compile-time checks.
type CounterUpdate struct {
	CounterID uint32 `json:"counter_id"` // Manual tags
	NewCount  int32  `json:"new_count"`  // Manual tags
	Timestamp uint64 `json:"timestamp"`  // Manual tags
}

func fibonacci(n int32) int32 {
	if n <= 1 {
		return n
	}
	return fibonacci(n-1) + fibonacci(n-2)
}

func main() {
	fmt.Println("Go counter service: manual locks, no supervision, no wire types")
	fmt.Println("Lines of domain logic: 80+")
	fmt.Println("External dependencies: sync, encoding/json (or protobuf)")
	fmt.Println("Runtime overhead: GC pauses, goroutine scheduler")
	result := fibonacci(10)
	fmt.Printf("fibonacci(10) = %d\n", result)
}

File: examples/comparison/counter_service.rs
// Rust: Counter Service
// Same problem as counter_service.hew but in Rust.
// Notice: Arc<Mutex<>> everywhere, no actors, no supervision,
// no wire types without serde + format crate.
use std::sync::{Arc, Mutex};

// Rust has no actor model. You use Arc<Mutex<>> for shared state.
// Safe, but verbose and easy to deadlock.

struct Counter {
    count: i32,
}

impl Counter {
    fn new() -> Self {
        Counter { count: 0 }
    }

    fn increment(&mut self, amount: i32) -> i32 {
        self.count += amount;
        self.count
    }

    fn get_count(&self) -> i32 {
        self.count
    }
}

// To share across threads, you need Arc<Mutex<Counter>>
type SharedCounter = Arc<Mutex<Counter>>;

fn new_shared_counter() -> SharedCounter {
    Arc::new(Mutex::new(Counter::new()))
}

fn shared_increment(counter: &SharedCounter, amount: i32) -> i32 {
    let mut c = counter.lock().unwrap(); // Can panic on poison
    c.increment(amount)
}

// Rust has no supervision trees. You build your own.
struct CounterPool {
    counters: Vec<SharedCounter>,
    max_restarts: i32,
    restart_count: i32,
}

impl CounterPool {
    fn new(size: usize) -> Self {
        let counters = (0..size).map(|_| new_shared_counter()).collect();
        CounterPool {
            counters,
            max_restarts: 5,
            restart_count: 0,
        }
    }

    // No language support for restart strategies.
    fn restart_child(&mut self, idx: usize) {
        self.restart_count += 1;
        if self.restart_count > self.max_restarts {
            eprintln!("Exceeded restart limit");
            return;
        }
        self.counters[idx] = new_shared_counter();
    }
}

// Rust has no wire types. You need serde + a format crate (bincode, etc.)
// Plus derive macros. No compile-time schema evolution checking.
// #[derive(Serialize, Deserialize)]  // Needs serde dependency
struct CounterUpdate {
    counter_id: u32,
    new_count: i32,
    timestamp: u64,
}

fn fibonacci(n: i32) -> i32 {
    if n <= 1 {
        n
    } else {
        fibonacci(n - 1) + fibonacci(n - 2)
    }
}

fn main() {
    println!("Rust counter service: Arc<Mutex<>>, no supervision, no wire types");
    println!("Lines of domain logic: 90+");
    println!("External dependencies: serde, bincode (or similar) for wire types");
    println!("Runtime overhead: native binary, no GC (advantage shared with Hew)");
    let result = fibonacci(10);
    println!("fibonacci(10) = {}", result);
}
